python/ray/tune/utils/file_transfer.py

```diff
@@ -0,0 +1,236 @@
+                # Add files where our information differs
+                for file in files:
+                    key = os.path.join(rel_root, file)
+                    stat = os.lstat(os.path.join(source_dir, key))
+                    file_stat = stat.st_mtime, stat.st_size
+                    if key not in files_stats or file_stat != files_stats[key]:
+                        tar.add(os.path.join(source_dir, key), arcname=key)
+
+    return stream
+
+
+@ray.remote
+class _PackActor:
+    """Actor wrapping around a packing job.
+
+    This actor is used for chunking the packed data into smaller chunks that
+    can be transferred via the object store more efficiently.
+
+    The actor will start packing the directory when initialized, and separate
+    chunks can be received by calling the remote ``next()`` task.
+
+    Args:
+        source_dir: Path to local directory to pack into tarfile.
+        files_stats: Dict of relative filenames mapping to a tuple of
+            (mtime, filesize). Only files that differ from these stats
+            will be packed.
+        chunk_size: Cut bytes stream into chunks of this size in bytes.
+    """
+
+    def __init__(
+        self,
+        source_dir: str,
+        files_stats: Optional[Dict[str, Tuple[float, int]]] = None,
+        chunk_size: int = DEFAULT_CHUNK_SIZE,
+    ):
+        self.stream = _pack_dir(source_dir=source_dir, files_stats=files_stats)
+        self.chunk_size = chunk_size
+        self.iter = None
+
+    def get_full_data(self) -> bytes:
+        return self.stream.getvalue()
+
+    def _chunk_generator(self) -> Generator[None, bytes, None]:
+        self.stream.seek(0)
+        data = self.stream.read(self.chunk_size)
+        while data:
+            yield data
+            data = self.stream.read(self.chunk_size)
+
+    def next(self) -> Optional[bytes]:
+        if not self.iter:
+            self.iter = iter(self._chunk_generator())
+        try:
+            return next(self.iter)
+        except StopIteration:
+            return None
+
+
+def _iter_remote(actor: ray.ActorID) -> Generator[None, bytes, None]:
+    """Iterate over actor task and return as generator."""
+    while True:
+        try:
+            buffer = ray.get(actor.next.remote())
+            if buffer is None:
+                return
+            yield buffer
+        except RayTaskError:
+            return None
+
+
+def _unpack_dir(stream: io.BytesIO, target_dir: str):
+    """Unpack tarfile stream into target directory."""
+    stream.seek(0)
+    with tarfile.open(fileobj=stream) as tar:
+        tar.extractall(target_dir)
+
+
+@ray.remote
+def _unpack_from_actor(pack_actor: ray.ActorID, target_dir: str):
+    """Iterate over chunks received from pack actor and unpack."""
+    stream = io.BytesIO()
+    for buffer in _iter_remote(pack_actor):
+        stream.write(buffer)
+    _unpack_dir(stream, target_dir=target_dir)
+
+
+@ray.remote
+def _delete_path(target_path: str) -> bool:
```

Comment:
nit: This seems a bit odd to decorate with `@ray.remote` as the default functionality (I think the caller should always be explicit about which node its being executed on). Maybe we can just wrap this function with `ray.remote()` directly in `delete_on_node` instead?

Same for `_unpack_from_actor`

Reply:
See below, plus these are private functions only used in this file. We're not calling them in a non-remote way, we just decorate them to avoid exporting the same function multiple times. We can have `_remote_unpack_from_actor` but would never use `_unpack_from_actor` directly, so there's no need for that